[SPARK-17455][MLlib] Improve PAVA implementation in IsotonicRegression

neggert opened the pull request:
## What changes were proposed in this pull request?

New implementation of the Pool Adjacent Violators Algorithm (PAVA) in mllib.IsotonicRegression, which used under the hood by ml.regression.IsotonicRegression. The previous implementation could have factorial complexity in the worst case. This implementation, which closely follows those in scikit-learn and the R `iso` package, runs in quadratic time in the worst case.
## How was this patch tested?

Existing unit tests in both `mllib` and `ml` passed before and after this patch. Scaling properties were tested by running the `poolAdjacentViolators` method in [scala-benchmarking-template](https://github.com/sirthias/scala-benchmarking-template) with the input generated by

``` scala
val x = (1 to length).toArray.map(_.toDouble)
val y = x.reverse.zipWithIndex.map{ case (yi, i) => if (i % 2 == 1) yi - 1.5 else yi}
val w = Array.fill(length)(1d)

val input: Array[(Double, Double, Double)] = (y zip x zip w) map{ case ((y, x), w) => (y, x, w)}
```

Before this patch:

| Input Length | Time (us) |
| --: | --: |
| 100 | 1.35 |
| 200 | 3.14 |
| 400 | 116.10 |
| 800 | 2134225.90 |

After this patch:

| Input Length | Time (us) |
| --: | --: |
| 100 | 1.25 |
| 200 | 2.53 |
| 400 | 5.86 |
| 800 | 10.55 |

Benchmarking was also performed with randomly-generated y values, with similar results.

neggert committed:
New PAVA implementation

SparkQA commented:
**[Test build #65117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65117/consoleFull)** for PR 15018 at commit [`5608ae1`](https://github.com/apache/spark/commit/5608ae1f4c9089dd3a299eef14af59bed6c96192).

SparkQA commented:
**[Test build #65117 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65117/consoleFull)** for PR 15018 at commit [`5608ae1`](https://github.com/apache/spark/commit/5608ae1f4c9089dd3a299eef14af59bed6c96192).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65117/
Test PASSed.

neggert commented:
Could someone please take a look at this? @mengxr maybe? Looks like you merged the initial implementation.

neggert commented:
Anyone? @srowen?

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -344,27 +344,30 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+
+        // Find next monotonicity violating sequence, if any.
+        while (j < n && input(j)._1 >= input(j + 1)._1) {
+          j = j + 1
```
Nit: `j += 1` and likewise elsewhere.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -344,27 +344,30 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+        var j = i
+
+        // Find next monotonicity violating sequence, if any.
+        while (j < n && input(j)._1 >= input(j + 1)._1) {
```
The condition changed to >= here on purpose right? you're now looking for a region that's non-increasing, not just strictly decreasing. Just checking.

jkbradley replied:
That should be correct.  Lemma 1 from "A Fast Scaling Algorithm for Minimizing Separable Convex Functions Subject to Chain Constraints"

viirya replied:
I think the original one i.e., `input(j)._1 > input(j + 1)._1` is correct. Here it is going to select out-of-order blocks.

Quoted from the paper:
> We refer to two blocks [p, q] and [q + 1, r] as consecutive. We refer to two consecutive blocks [p, q] and [q +1, r] as in-order if  theta_pq <= theta_q+1, r and out-of-order otherwise.

LEMMA 1 is pointing the how a merged block is also a single-valued block.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -344,27 +344,30 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
-        // and check if pooling caused monotonicity violation in previously processed points.
-        while (i >= 0 && input(i)._1 > input(i + 1)._1) {
+        // Pool the violating sequence with the data point preceding it
+        if (input(i)._1 != input(j)._1) {
```
This condition is "if there was any decrease, not just == from i to j" right? and `pool` makes everything in [i,j] non-violating.

Comments from the previous code note that this could introduce a violation -- is this still possible? it wasn't obvious why it can't happen now.

neggert replied:
Correct. Any violations that are introduced will be fixed in the next iteration.

srowen reviewed (changes requested):
More than anything it would be great to add a couple simple unit tests of this method to prove it still works in key cases.

jkbradley commented:
Thanks for pinging about this---it looks important!

One request: Do you know if there is a resource with a clearer statement of this precise algorithm for the L2 norm?  That'd be nice to have.  The best I found so far was "A Fast Scaling Algorithm for Minimizing Separable Convex Functions Subject to Chain Constraints"

viirya commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -344,27 +344,30 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
-        // Otherwise pool the violating sequence
-        // and check if pooling caused monotonicity violation in previously processed points.
-        while (i >= 0 && input(i)._1 > input(i + 1)._1) {
+        // Pool the violating sequence with the data point preceding it
```
Is this comment correct? Looks like you pool the violating sequence [i, j] only. The preceding data points are checked for pooling in next outer loop.

SparkQA commented:
**[Test build #70134 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70134/consoleFull)** for PR 15018 at commit [`8d16043`](https://github.com/apache/spark/commit/8d16043b50629f2dc2f96c124189d284c833ddc0).

neggert commented:
Given that the scikit-learn implementation I based this on has changed (https://github.com/scikit-learn/scikit-learn/pull/7444) since I submitted the PR, and @jkbradley has pointed out a reference I hadn't seen, I'd like to take a little bit of time to see if there's yet a better algorithm.

This one is an improvement over what we had, but as long as we're changing it, we may as well get it right.

neggert committed:
Style and comments

SparkQA commented:
**[Test build #70136 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70136/consoleFull)** for PR 15018 at commit [`9cfa5ed`](https://github.com/apache/spark/commit/9cfa5ed57cbedcb5189d612c74f5c81c4cc0cc0b).

SparkQA commented:
**[Test build #70134 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70134/consoleFull)** for PR 15018 at commit [`8d16043`](https://github.com/apache/spark/commit/8d16043b50629f2dc2f96c124189d284c833ddc0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70134/
Test PASSed.

SparkQA commented:
**[Test build #70136 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70136/consoleFull)** for PR 15018 at commit [`9cfa5ed`](https://github.com/apache/spark/commit/9cfa5ed57cbedcb5189d612c74f5c81c4cc0cc0b).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70136/
Test FAILed.

neggert commented:
Looks like the new scikit-learn implementation suffers from a similar problem to the one that the original Spark implementation had. I left them a note pointing it out.

Right now I'm working on implementing what's in "Minimizing Separable Convex Functions Subject to Simple Chain Constraints", by Best, Chakravarti, and Ubhaya. Running down the referecence trail, it looks like the original algorithm is in "Projections onto Order Simplexes" by Grotzinger and Witzgall, and this particular formulation of it is originally in "Active set algorithms for isotonic regression; A unifying framework" by Best and Chakravarti.

srowen commented:
If it's correct and faster, I tend to favor merging this. The code looks OK to my understanding and passes existing tests.

neggert commented:
Found another input that triggers non-polynomial time with the code in this PR. I'm again borrowing from scikit-learn. I think this is the case they found that led them to re-write their implementation.

```
    val y = ((0 until length) ++ (-(length - 1) until length) ++ (-(length - 1) to 0)).toArray.map(_.toDouble)
    val x = (1 to y.length).toArray.map(_.toDouble)
```

| Input Length | Time (ns) |
| --: | --: |
| 40 | 2059 |
| 80 | 4604 |
| 160 | 1974269 |
| 320 | 3246603433 |

I'm now working on implementing what's described in the Best papers. This should give O(n), even in the worst case. 

Should I close this and open a new PR with the new algorithm, or just add it here and you can squash when you merge?

srowen commented:
Just add commits here.

neggert committed:
Remove test with 0 weights

The isotonic regression problem doesn't even have a unique solution
when there are 0 weights.

neggert committed:
Re-write PAV algorithm again.

This version closely follows the algorithm in referenced literature
and should guarantee O(n) complexity, even in the worst case.

SparkQA commented:
**[Test build #70214 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70214/testReport)** for PR 15018 at commit [`61d383f`](https://github.com/apache/spark/commit/61d383fd65d3a4e99938931483e6636025893039).

SparkQA commented:
**[Test build #70214 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70214/testReport)** for PR 15018 at commit [`61d383f`](https://github.com/apache/spark/commit/61d383fd65d3a4e99938931483e6636025893039).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70214/
Test FAILed.

neggert commented:
Alright, the PAV algorithm has been completely re-written to follow what's outlined in "Minimizing Separable Convex Functions Subject to Simple Chain Constraints". I've tested it with a bunch of different inputs that caused previous version of this algorithm to go non-polynomial. It stays linear for all of them. I will note that it's slightly slower on very small datasets (< 5000 points or so), but those still finish in less than a millisecond on my laptop, so I'm not too concerned.

The only caveat is that there was one test that I couldn't get passing, so I removed it. It involved passing input data with 0 weights. I'd argue that the isotonic regression problem doesn't even have a unique solution in that case, so we shouldn't support it. Still, it is a slight change in behavior.

Looks like Jenkins is pointing out some style issues, so I'll get to work on those.

neggert committed:
Style and comments

SparkQA commented:
**[Test build #70216 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70216/testReport)** for PR 15018 at commit [`e420fb3`](https://github.com/apache/spark/commit/e420fb353c079d8053695ad2b6cf13a52fdf2e05).

SparkQA commented:
**[Test build #70216 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70216/testReport)** for PR 15018 at commit [`e420fb3`](https://github.com/apache/spark/commit/e420fb353c079d8053695ad2b6cf13a52fdf2e05).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70216/
Test FAILed.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -328,74 +336,68 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+    while (nextBlock(i) < input.length) {
+      if (average(i) >= average(nextBlock(i))) {
+        merge(i, nextBlock(i))
+        while((i > 0) && (average(prevBlock(i)) >= average(i))) {
```
Super nit: space before `while` here and below. You can probably nix the extra parens around each of the two terms, but no big deal

srowen reviewed (commented):
Modulo a few minor comments, the code at least looks clean. If the tests pass, seems OK.

neggert committed:
Style fixes

SparkQA commented:
**[Test build #70363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70363/testReport)** for PR 15018 at commit [`8762531`](https://github.com/apache/spark/commit/8762531b9c70e21ce047393b252fefe2ed134036).

SparkQA commented:
**[Test build #70363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70363/testReport)** for PR 15018 at commit [`8762531`](https://github.com/apache/spark/commit/8762531b9c70e21ce047393b252fefe2ed134036).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70363/
Test FAILed.

SparkQA commented:
**[Test build #70364 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70364/testReport)** for PR 15018 at commit [`d19647b`](https://github.com/apache/spark/commit/d19647b230f3350432b67a5035822cb04d529c9d).

SparkQA commented:
**[Test build #70364 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70364/testReport)** for PR 15018 at commit [`d19647b`](https://github.com/apache/spark/commit/d19647b230f3350432b67a5035822cb04d529c9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70364/
Test PASSed.

neggert commented:
@srowen Addressed your comments and fixed some style issues.

Some updated timings:

Alternating decreasing input

    val x = (1 to length).toArray.map(_.toDouble)
    val y = x.reverse.zipWithIndex.map{ case (yi, i) => if (i % 2 == 1) yi - 1.5 else yi}

Before: 

| Input Length | Time (us) |
|-------------:|----------:|
| 100 | 1.4 |
| 200 | 3.0 |
| 400 | 101.8 |
| 800 | 2,010,511.7 |

After:

| Input Length | Time (us) |
|-------------:|----------:|
| 100 | 2.5 |
| 200 | 5.8 |
| 400 | 10.1 |
| 800 | 20.7 |

Pathological input from scikit-learn benchmarks:

    val y= ((0 until length) ++ (-(length - 1) until length) ++ (-(length - 1) to 0)).toArray.map(_.toDouble)
    val x = (1 to y.length).toArray.map(_.toDouble)

Before:

| Input Length | Time (us) |
|-------------:|----------:|
| 40 | 2.2 |
| 80 | 4.7 |
| 160 | 2140.0 |
| 320 | 3,076,508.1 |

After:

| Input Length | Time (us) |
|-------------:|----------:|
| 40 | 5.3 |
| 80 | 10.4 |
| 160 | 21.3 |
| 320 | 41.9 |

viirya commented:
For the zero-weight values, can we do similar to scikit-learn to remove zero-weight values, like https://github.com/amueller/scikit-learn/commit/2415100f79293bbbf52c12c36d63a6cf602cf3c4

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -328,74 +336,69 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
-      val weight = poolSubArray.map(_._3).sum
+    // Keeps track of the start and end indices of the blocks. blockBounds(start) gives the
+    // index of the end of the block and blockBounds(end) gives the index of the start of the
+    // block. Entries that are not the start or end of the block are meaningless. The idea is that
```
I'm still not sure about this comment -- how can `blockBounds(x)` be both the start and end of a block? the implementation below is identical.

neggert replied:
It relies on knowing ahead of time wether `x` is the start or end index<sup>1</sup>. If it's a start index, `blockBounds(x)` gives the ending index of that block. If it's an end index, `blockBounds(x)` gives the starting index of the block. So yes, the implementations of `blockStart` and `blockEnd` are identical. I just have two different functions because it makes the code more readable.

Maybe the comment from scikit-learn (where I borrowed this idea from) explains it better? (their `target` = my `blockBounds`)

> target describes a list of blocks.  At any time, if [i..j] (inclusive) is
> an active block, then blockBounds[i] := j and target[j] := i.

The trick is just in maintaining the array so that the above property is always true. At initialization, it's trivially true because all blocks have only one element, and `blockBounds(x)` = `x`. After initialization, `blockBounds` is only modified by the merge function, which is set up to modify `blockBounds` so that this property is preserved, then return the starting index of the newly-merged block.

This is admittedly a bit tricky, but it's a lot faster than the implementation I did where created a doubly-linked list of `Block` case classes. I'm open to suggestions on how to explain it better.

<sup>1</sup> You could actually figure out whether you have a start or an end index by comparing `blockBounds(x)` to `x`. The lesser value will be the start index.

neggert commented:
@viirya Better to remove them, or throw an error? Personally, I'd rather be alerted that I'm passing invalid input, rather than have it "fixed" for me.

viirya commented:
@neggert I am fine for throwing an error.

srowen commented:
@neggert I think this is looking good except for the question about zero weights. If throwing an error is OK to you go for it. The test should be restored, but be updated to expect an exception. Other than that, more tests always help, but there is some OK coverage already

srowen commented:
Ping @neggert

neggert commented:
Sorry, I've been on vacation. I'll make the requested changes in the next day or two.

neggert committed:
Revert "Remove test with 0 weights"

This reverts commit 07fe1d9705ca41c26e1114b38a80d30703779cac.

neggert committed:
Update comments

neggert committed:
Throw exception when given 0 weights

SparkQA commented:
**[Test build #70831 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70831/testReport)** for PR 15018 at commit [`c65e263`](https://github.com/apache/spark/commit/c65e2639741fb3e4717c2f16922d3b10c9f6e2e7).

SparkQA commented:
**[Test build #70831 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70831/testReport)** for PR 15018 at commit [`c65e263`](https://github.com/apache/spark/commit/c65e2639741fb3e4717c2f16922d3b10c9f6e2e7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70831/
Test FAILed.

neggert committed:
Style

SparkQA commented:
**[Test build #70833 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70833/testReport)** for PR 15018 at commit [`da1ac1b`](https://github.com/apache/spark/commit/da1ac1b650c52428633a7efa97f3be7d2dbfae57).

SparkQA commented:
**[Test build #70833 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70833/testReport)** for PR 15018 at commit [`da1ac1b`](https://github.com/apache/spark/commit/da1ac1b650c52428633a7efa97f3be7d2dbfae57).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70833/
Test FAILed.

neggert committed:
Style

SparkQA commented:
**[Test build #70836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70836/testReport)** for PR 15018 at commit [`47d5b4f`](https://github.com/apache/spark/commit/47d5b4fc58bc40fcec2565185d4a4068219ebb47).

SparkQA commented:
**[Test build #70836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70836/testReport)** for PR 15018 at commit [`47d5b4f`](https://github.com/apache/spark/commit/47d5b4fc58bc40fcec2565185d4a4068219ebb47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70836/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -328,74 +336,81 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+    // gives the values for the block. Entries that are not at the start of a block
+    // are meaningless.
+    val weights: Array[(Double, Double)] = input.map {
+      case (_, _, weight) if weight == 0d =>
```
I'd always write 0.0 instead of 0d for clarity. I also think we want an `IllegalArgumentException`, so maybe:

```
val weights = input.map { case (y, _, weight) =>
  require(weight > 0.0)
  (weight, weight * y)
}
```

srowen reviewed (changes requested):
Just a minor style request

neggert committed:
Use require instead of throwing SparkException

neggert commented:
Done. It was surprisingly tricky to test, since the exception is thrown in the executor.

SparkQA commented:
**[Test build #70886 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70886/testReport)** for PR 15018 at commit [`a119423`](https://github.com/apache/spark/commit/a119423fa769bb6be9cc27e455a82d04360e03b3).

SparkQA commented:
**[Test build #70886 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70886/testReport)** for PR 15018 at commit [`a119423`](https://github.com/apache/spark/commit/a119423fa769bb6be9cc27e455a82d04360e03b3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70886/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -328,74 +336,80 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+    // gives the values for the block. Entries that are not at the start of a block
+    // are meaningless.
+    val weights: Array[(Double, Double)] = input.map { case (y, _, weight) =>
+      require(weight != 0.0)
```
Are negative weights OK? (you don't need a type on `weight`, but hardly matters)

neggert replied:
It's a weird thing to do, but I don't think there's any reason we can't support it. The problem still has a unique solution.

jkbradley replied:
Please add an informative error message.

srowen reviewed (approved):
OK, seems reasonable to me. Any other thoughts @viirya ?

viirya commented:
SGTM. See if @jkbradley has more comment?

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -328,74 +336,80 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+    // Merge two adjacent blocks, updating blockBounds and weights to reflect the merge
+    // Return the start index of the merged block
+    def merge(block1: Int, block2: Int): Int = {
+      assert(blockEnd(block1) + 1 == block2, "attempting to merge non-consecutive blocks")
```
Please make the error message more informative:
* Make it clear that this indicates an internal bug within IsotonicRegression
* State the invalid values

jkbradley reviewed (commented):
Thanks for pinging me.  The updated implementation looks nice, but I do have comments about weights.

I'd like to disallow negative weights.  It's quite safe to assume that weights are non-negative in ML, and I've never seen negative weights used in isotonic regression.
* Q: I saw mention of negative weights above, but the Best and Chakravarti (1990) paper referenced explicitly requires strictly positive weights.  Does one of the other papers talk about negative weights?
* Btw, with negative weights, it would be easy to have undefined solutions.  Imagine the first element in the input having the smallest value and a negative weight.  A "correct" output would be infinite for that value and 0 elsewhere, giving a minimum objective value of NegativeInfinity.

I'd also prefer to allow zero weights to avoid breaking the API and since the model is fairly well-defined even with zero weights, except for edge cases at extreme feature values.  Removing the zero-weight rows should provide reasonable solutions, though we'd need to assert that the whole dataset had non-zero weight.

How does this sound?

neggert commented:
Your changes sound fine. I don't have strong feelings one way or another, although I think we should at least throw a warning if we're discarding points. I do want to point out that disallowing negative weights also breaks the API, since they were previously supported (there's even a test for them).

SparkQA commented:
**[Test build #71087 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71087/testReport)** for PR 15018 at commit [`e33a308`](https://github.com/apache/spark/commit/e33a3083fc6437bcf5149103e71646f355e84578).

SparkQA commented:
**[Test build #71087 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71087/testReport)** for PR 15018 at commit [`e33a308`](https://github.com/apache/spark/commit/e33a3083fc6437bcf5149103e71646f355e84578).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71087/
Test FAILed.

neggert committed:
Automatically remove 0-weight points and prohibit negative weights

SparkQA commented:
**[Test build #71092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71092/testReport)** for PR 15018 at commit [`83480e7`](https://github.com/apache/spark/commit/83480e7ac33d9cf066245f2e8206ed6d765bafb4).

SparkQA commented:
**[Test build #71092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71092/testReport)** for PR 15018 at commit [`83480e7`](https://github.com/apache/spark/commit/83480e7ac33d9cf066245f2e8206ed6d765bafb4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71092/
Test PASSed.

viirya commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -312,10 +313,19 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+   * functions subject to simple chain constraints." SIAM Journal on Optimization 10.3 (2000):
+   * 658-672.
    *
    * @param input Input data of tuples (label, feature, weight).
```
Can you update the param doc to indicate that now negative weights are prohibited?

viirya commented:
LGTM

neggert committed:
Update docstring

SparkQA commented:
**[Test build #71271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71271/testReport)** for PR 15018 at commit [`132d6a6`](https://github.com/apache/spark/commit/132d6a691a0a5f65b6d2cd84d031883a6e917282).

SparkQA commented:
**[Test build #71271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71271/testReport)** for PR 15018 at commit [`132d6a6`](https://github.com/apache/spark/commit/132d6a691a0a5f65b6d2cd84d031883a6e917282).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71271/
Test PASSed.

jkbradley commented:
@neggert I'll ask @mengxr about the negative weights since he oversaw the original work here.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -312,90 +313,120 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+    val cleanInput = input.flatMap{ case (y, x, weight) =>
+      require(weight >= 0.0, "weights must be non-negative")
+      if (weight == 0.0) {
+        logWarning(s"Dropping zero-weight point ($y, $x, $weight)")
```
I disagree about logging for zero-weight points.  If an instance has zero weight, then it should be ignored, and that's not a problem.

If we really want to log this, then let's do it once per dataset, not once per row, which could make logs explode.  I'd also say we should lower the logging level to logInfo.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -312,90 +313,120 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
-    if (input.isEmpty) {
-      return Array.empty
+    val cleanInput = input.flatMap{ case (y, x, weight) =>
+      require(weight >= 0.0, "weights must be non-negative")
```
State invalid y,x,weight.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -312,90 +313,120 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
 
-    if (input.isEmpty) {
-      return Array.empty
+    val cleanInput = input.flatMap{ case (y, x, weight) =>
```
filter would be simpler than flatMap

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala:
```diff
@@ -312,90 +313,120 @@ class IsotonicRegression private (private var isotonic: Boolean) extends Seriali
+      assert(
+        blockEnd(block1) + 1 == block2,
+        s"Attempting to merge non-consecutive blocks [${block1}, ${blockEnd(block1)}]" +
+        s" and [${block2}, ${blockEnd(block2)}. This is likely a bug in the isotonic regression" +
```
missing "]"

jkbradley reviewed (commented):
Just a few comments, other than checking with @mengxr about negative weights.

jkbradley commented:
@zapletal-martin  Pinging since you wrote the original PR: There's discussion here about whether IsotonicRegression should support negative weights.  Is there a good reason to?  I haven't seen papers which allow negative weights, though I've only looked at a few.  Thanks!

neggert commented:
My thought was that there's nothing in the computation that prevents it (unlike 0 weights, which cause NaNs). Why not err on the side of maximum flexibility to the end user? Just because there's no use case for it now doesn't mean someone won't find one in the future.

Just my 2 cents. I don't feel strongly about it and will defer to commiters.

neggert committed:
Minor clean-up and remove warning for 0 weight points

SparkQA commented:
**[Test build #71329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71329/testReport)** for PR 15018 at commit [`0aeca09`](https://github.com/apache/spark/commit/0aeca09c0ea234d8a61664f2827abd80d7850c5b).

neggert commented:
Looks like the build completed successfully. Somehow Jenkins didn't post the message.

jkbradley commented:
Thanks for the updates!  This LGTM, except for deciding about negative weights.

Responding to your comment above, negative weights are just as problematic as 0 weights.  See my comment above "Btw, with negative weights, ..."

I spoke with @mengxr offline, and he does not recall why the original work supported negative weights.  I'll wait a few days before merging this to see if @zapletal-martin can check out the issue.

srowen commented:
@jkbradley I'll merge if there are no other thoughts here. Seems OK to me too

jkbradley commented:
Sounds good.  I'll run fresh tests before merging to be safe though.

SparkQA commented:
**[Test build #3547 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3547/testReport)** for PR 15018 at commit [`0aeca09`](https://github.com/apache/spark/commit/0aeca09c0ea234d8a61664f2827abd80d7850c5b).

SparkQA commented:
**[Test build #3547 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3547/testReport)** for PR 15018 at commit [`0aeca09`](https://github.com/apache/spark/commit/0aeca09c0ea234d8a61664f2827abd80d7850c5b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
Merging with master
Thanks!
